library/alloc/src/vec/into_iter.rs

```diff
@@ -132,7 +134,9 @@ impl<T, A: Allocator> IntoIter<T, A> {
 
     /// Forgets to Drop the remaining elements while still allowing the backing allocation to be freed.
     pub(crate) fn forget_remaining_elements(&mut self) {
-        self.ptr = self.end;
+        // For th ZST case, it is crucial that we mutate `end` here, not `ptr`.
+        // `ptr` must stay aligned, while `end` may be unaligned.
+        self.end = self.ptr;
```

Comment:
FWIW I don't think we have a test for this (with a ZST that needs alignment)

Reply:
Wouldn't it be better to add the tests to std instead of the miri testsuite then?

Reply:
I'm not sure how this test would be written in std in a way that would actually test the behavior except under miri.

Reply:
Maybe by adding an unsafe precondition debug assert to drop_in_place?

Reply:
I would not be opposed to that, but I don't think it needs to be done in this PR.